File: fn/src/main/java/org/sqlbuilder/fn/objects/Frame.java
package org.sqlbuilder.fn.objects;

import org.sqlbuilder.common.Insertable;
import org.sqlbuilder.common.Utils;
import org.sqlbuilder.fn.HasID;
import org.sqlbuilder.fn.collectors.FnFrameXmlProcessor;
import org.xml.sax.SAXException;

import java.io.IOException;
import java.util.Comparator;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;

import javax.xml.parsers.ParserConfigurationException;

import edu.berkeley.icsi.framenet.FrameDocument;

public class Frame implements HasID, Insertable
{
	public static final Set<Frame> SET = new HashSet<>();

	private static final FnFrameXmlProcessor definitionProcessor = new FnFrameXmlProcessor();

	public final int frameid;

	public final String name;

	public final String definition;

	public static Frame make(final FrameDocument.Frame frame) throws IOException, SAXException, ParserConfigurationException
	{
		var f = new Frame(frame);
		SET.add(f);
		return f;
	}

	private Frame(final FrameDocument.Frame frame) throws IOException, SAXException, ParserConfigurationException
	{
		this.frameid = frame.getID();
		this.name = frame.getName();
		try
		{
			this.definition = Frame.definitionProcessor.process(frame.getDefinition());
		}
		catch (ParserConfigurationException | SAXException | IOException e)
		{
			System.err.println(frame.getDefinition());
			throw e;
		}
	}

	// A C C E S S

	public long getID()
	{
		return frameid;
	}

	public String getName()
	{
		return name;
	}

	public String getDefinition()
	{
		return definition;
	}

	// I D E N T I T Y

	@Override
	public boolean equals(final Object o)
	{
		if (this == o)
		{
			return true;
		}
		if (o == null || getClass() != o.getClass())
		{
			return false;
		}
		Frame that = (Frame) o;
		return frameid == that.frameid;
	}

	@Override
	public int hashCode()
	{
		return Objects.hash(frameid);
	}

	// O R D E R

	public static final Comparator<Frame> COMPARATOR = Comparator.comparing(Frame::getName).thenComparing(Frame::getID);

	// I N S E R T

	@Override
	public String dataRow()
	{
		return String.format("%d,'%s','%s'", //
				frameid, //
				Utils.escape(name), //
				Utils.escape(definition));
		// frame.getCDate()
		// frame.getCBy()
	}

	// T O S T R I N G

	@Override
	public String toString()
	{
		return String.format("[FRAME id=%s name=%s]", frameid, name);
	}
}

File: fn/src/main/java/org/sqlbuilder/fn/objects/Frame.kt
package org.sqlbuilder.fn.objects

import edu.berkeley.icsi.framenet.FrameDocument
import org.sqlbuilder.common.HasID
import org.sqlbuilder.common.Insertable
import org.sqlbuilder.common.Utils.escape
import org.sqlbuilder.fn.collectors.FnFrameXmlProcessor
import org.xml.sax.SAXException
import java.io.IOException
import java.util.*
import javax.xml.parsers.ParserConfigurationException

class Frame private constructor(
    frame: FrameDocument.Frame,
) : HasID, Insertable {

    val iD: Int = frame.getID()

    val name: String = frame.getName()

    val definition: String

    init {
        try {
            definition = definitionProcessor.process(frame.getDefinition())
        } catch (e: ParserConfigurationException) {
            System.err.println(frame.getDefinition())
            throw e
        } catch (e: SAXException) {
            System.err.println(frame.getDefinition())
            throw e
        } catch (e: IOException) {
            System.err.println(frame.getDefinition())
            throw e
        }
    }

    // I D E N T I T Y

    override fun equals(o: Any?): Boolean {
        if (this === o) {
            return true
        }
        if (o == null || javaClass != o.javaClass) {
            return false
        }
        val that = o as Frame
        return iD == that.iD
    }

    override fun hashCode(): Int {
        return Objects.hash(iD)
    }

    // I N S E R T

    override fun dataRow(): String {
        return "$iD,'${escape(name)}','${escape(definition)}'"
    }

    // T O S T R I N G

    override fun toString(): String {
        return "[FRAME id=$iD name=$name]"
    }

    companion object {

        val COMPARATOR: Comparator<Frame> = Comparator
            .comparing<Frame, String> { it.name }
            .thenComparing<Int> { it.iD }

        val SET = HashSet<Frame>()

        private val definitionProcessor = FnFrameXmlProcessor()

        @Throws(IOException::class, SAXException::class, ParserConfigurationException::class)
        fun make(frame: FrameDocument.Frame): Frame {
            val f = Frame(frame)
            SET.add(f)
            return f
        }
    }
}
